R/function_options.R

```diff
@@ -0,0 +1,69 @@
+extendr_function_config <- rlang::env(
+  known_options = tibble::tribble(
+    ~Name, ~Ptype,
+    "use_try_from", logical(),
+    "r_name", character(),
+    "mod_name", character(),
+  )
+)
+
+convert_function_options <- function(options, quiet) {
+  if (rlang::is_null(options) || rlang::is_empty(options)) {
+    return(tibble::tibble(Name = character(), RustValue = character()))
+  }
+
+  if (!rlang::is_list(options) || !rlang::is_named(options)) {
+    cli::cli_abort("Extendr function options should be either a named {.code list()} or {.code NULL}.")
+  }
+
+  options_table <- tibble::tibble(Name = rlang::names2(options), Value = unname(options)) %>%
+    dplyr::left_join(extendr_function_config$known_options, by = "Name") %>%
+    dplyr::mutate(
+      Value = purrr::pmap(
+        list(.data$Value, .data$Ptype, .data$Name),
+        ~ if (rlang::is_null(..2)) ..1 else vctrs::vec_cast(..1, ..2, x_arg = ..3)
+      ),
+    )
+
+  unknown_options <- options_table %>%
+    dplyr::filter(purrr::map_lgl(.data$Ptype, rlang::is_null)) %>%
+    dplyr::pull(.data$Name)
+
+  invalid_options <- options_table %>%
+    dplyr::filter(
+      purrr::map_lgl(
+        .data$Value,
+        ~ rlang::is_null(.x) || !vctrs::vec_is(.x, size = 1L)
+      ) |
+        !is_valid_rust_name(.data$Name)
+    ) %>%
+    dplyr::pull(.data$Name)
+
+  if (length(invalid_options) > 0) {
+    cli::cli_abort(c(
+      "Found invalid {.code extendr} function option{?s}: {.val {invalid_options}}.",
+      "x" = "Option value should not be {.code NULL};",
+      "i" = "Option is expected to have scalar value;",
+      "i" = "Option name should be a valid rust identifier name."
```

Comment:
Am I right in thinking that these warnings will be printed even if they aren't relevant to the actual issue at hand?

Reply:
Yep. We already have an issue with this #253 , I am still thinking of how to do it best. I plan to first merge this and then revisit all of the warnings recently added.

Reply:
One way to do it might be to add error columns to the data frame. Like `df %>% mutate(invalid_id = ...) %>% mutate(non_scalar_value = ...) %>% mutate(is_invalid = invalid_id | non_scalar_value)`. So then you can query each error class and also the overall error status of each row when reporting back to the user.

Reply:
Yep, also an option. I do not want to over-engineer it, but still think that we need to inform users like 'hey, this looks suspicious'. A also think I need to pass something like `use_dev_extendr` to all of these methods so I could emit different warnings (see #253). The idea is that if released extendr is used but a user specifies an unknown option, this is likely an error, but if a dev version is used, we do not really *know* if this is a new option or an error, so the message should differ.